arrow-udf-deno/src/lib.rs

```diff
@@ -0,0 +1,811 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+use std::{cell::RefCell, collections::HashMap, rc::Rc, sync::Arc, task::Poll};
+
+use anyhow::{Context, Result};
+use arrow_array::{builder::Int32Builder, RecordBatch};
+use arrow_schema::{DataType, Field, Schema, SchemaRef};
+use arrow_udf_deno_runtime::deno_runtime;
+use futures::{Future, Stream, StreamExt, TryStreamExt};
+
+use crate::deno_arrow::get_jsvalue;
+
+mod deno_arrow;
+pub mod tokio_spawn_pinned;
+mod v8;
+mod values_future;
+
+thread_local! {
+    static THREAD_ISOLATE: Rc<RefCell<InternalRuntime>>  = Rc::new(RefCell::new( InternalRuntime::new() ));
+    static THREAD_RUNTIME: Arc<Runtime>  = Arc::new( Runtime::new_internal() );
```

Comment:
This is because each v8 can only run in one thread, it has to be created and used from the same thread, therefore it can not be send, the only exception is the isolate, since you can called from a different thread to cancel/interrupt the execution